### PR TITLE
feat: Support ecdsa and RSA keys (#270 with backwards compatibility)

### PR DESCRIPTION
**Pull request by asraa**
Please fill in the fields below to submit a pull request.  The more information that is provided, the better.

Fixes https://github.com/theupdateframework/go-tuf/issues/223
Release Notes: 
* breaking change: ECDSA public key verification now use TUF compliant PEM-encoding keys. For backwards compatibility with hex-encoded keys verification, client libraries must import `pkg/deprecated`
* feat: Adds support for ECDSA and RSA signers 
* feat: Repository CLI command allows a `--scheme` flag to specify key scheme.

**Types of changes**:

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Description of the changes being introduced by the pull request**:
* Defined new go types for KeyType, KeyScheme and HashAlgorithm, rather than just using strings, and fixed a few places where the key type and scheme had been mixed up.
* Added a new data.PKIXPublicKey type which abstracts away a lot of the boiler plate for marshalling/unmarshalling public keys.
* Completes support for the RSA signer and fixes public key encoding.

**Please verify and check that the pull request fulfills the following requirements**:

- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

cc @toby-jn: The only changes I made was adding deprecation support and updating your branch!

### CODE REVIEW

**Comment by trishankatdatadog** [comment]
So does this supersede #270 ?

**Comment by asraa** [comment]
> So does this supersede https://github.com/theupdateframework/go-tuf/pull/270 ?

Yes! It's all the changes there with backwards compatibility for just ECDSA verification: i.e. if a current root (sigstore) is using hex-encoded keys, then they will still be able to verify (no hex-encoded signers supported). Adds testing for making sure deprecation still verifies at the key and repo testing level.

**Comment by asraa** [comment]
I can also split out just the ECDSA changes. Hoping to get that in ASAP since the next sigstore TUF root to rotate those out relies on it.

**Comment by toby-jn** [comment]
Great thanks, LGTM. I did start trying to rebase my changes onto master over the weekend to make these updates, but there were some conflicts that I needed to go read up on and then real life got in the way, so thanks for taking this over.

**Comment by trishankatdatadog** [comment]
Thanks very much for all your hard work, Toby! Sorry I didn't get around to reviewing it in time: swamped by work, too...

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa.go:182]: It's not likely to be a large metadata file, but should we watch out for it?

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa.go:224]:
```suggestion
		return errors.New("unsupported ecdsa curve")
```

File: pkg/keys/ecdsa_test.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa_test.go:34]:
```suggestion
func (ECDSASuite) TestSignVerifyDeprecatedFormat(c *C) {
```

File: pkg/keys/rsa.go
Comment (trishankatdatadog) [inline pkg/keys/rsa.go:70]: Good riddance!

File: pkg/keys/rsa_test.go
Comment (trishankatdatadog) [inline pkg/keys/rsa_test.go:32]: Missing a few more tests, no, like signing and verifying? Would be nice to see the tests match the ecdsa ones.

File: repo.go
Comment (trishankatdatadog) [inline repo.go:323]: This is where I think things get confusing. I understand that key types and schemes are distinct, but if we are only going to support "rsa-pss-sha256", then we should use schemes instead of types here to make things sparkling clear.

**Review by trishankatdatadog — COMMENTED** [review]
Good stuff, thanks @asraa (and @toby-jn)!

Sorry for my late review, but a few comments...

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:43]: Guarding behind a flag would be great! I could imagine a build tag too that only adds the deprecated if the tag is enabled. But this API can be called as a library, and not as a compiled package. So maybe some options are:
* An environment variable, go-tuf specific: `GO_TUF_ENABLE_DEPRECATED_ECDSA`
* Adding an `init` function for deprecated keys into a separate package, and not including that into the default binary. Then only libraries would be able to use it on import.

File: cmd/tuf/gen_key.go
Comment (asraa) [reply cmd/tuf/gen_key.go:48]: Fixed, good catch! likely because err was not defined then.

File: repo.go
Comment (asraa) [reply repo.go:323]: I agree, changing to scheme.

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:182]: Since all signers implement this, I wonder if it should be implemented at a higher level, before the json becomes canonicalized bytes. I can file a follow-up issue. I'm hand-wavy on whether it matters: signing is never done by the client so it's not as much of a threat.

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [reply pkg/keys/ecdsa.go:43]: I like the second one, if it's not too much work.

Comment (asraa) [reply pkg/keys/ecdsa.go]: Me too! All set. I have to add a separate package `pkg/deprecated` so that the init function only runs when someone pull this in (as opposed to the keys package). Added a repo level test in there so that you can verify repository metadata with hex-encoded keys. Modified the `repo_test.go` so that it *fails* with the hex-encoded keys (same test, just failure)

**Comment by asraa** [comment]
@theupdateframework/go-tuf-maintainers could I please get one more review on this? really need this

**Comment by trishankatdatadog** [comment]
> @theupdateframework/go-tuf-maintainers could I please get one more review on this? really need this

Yes, sorry, was on PTO, will look tonight

**Comment by asraa** [comment]
No worries at all -- didn't realize that! Feel free to do it tomorrow, I hope you had some rest :)

File: pkg/deprecated/deprecated_repo_test.go
Comment (trishankatdatadog) [inline pkg/deprecated/deprecated_repo_test.go:77]: I'm confused. These don't look like they're failing?

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa.go:39]: Why are we not checking for an error here?

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa.go:87]: No checking for errors?

File: pkg/keys/rsa.go
Comment (trishankatdatadog) [inline pkg/keys/rsa.go:39]: No error checking?

File: pkg/keys/rsa.go
Comment (trishankatdatadog) [inline pkg/keys/rsa.go:84]: No error checking?

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [inline pkg/keys/ecdsa.go:88]: (BTW, side note: I'm beginning to think the TUF spec should also specify the encoding expect for public keys ahead of time, but that's a different story for another time.) @mnm678

File: pkg/keys/rsa.go
Comment (trishankatdatadog) [inline pkg/keys/rsa.go:78]: (Side note: no need to fix this now, and sticking to auto salt lengths for signing/verifying makes sense, but at some point, may make sense to encode this as part of the public key. See my related [Vault PR](https://github.com/hashicorp/vault/pull/16549).)

**Review by trishankatdatadog — COMMENTED** [review]
Looks great to me, thanks very much! Just a few comments...

File: cmd/tuf/gen_key.go
Comment (asraa) [reply cmd/tuf/gen_key.go:36]: Ohhh, I was operating under if `--scheme` isn't explicitly provided, then default ot Ed25519 (like existing behavior). I don't think people should always have to specify `--scheme`, same as how they don't have to specify `--expires`.

A compromise could be logging `Generated root ed25519 key with id xyz` just to inform users? Added.

File: pkg/deprecated/deprecated_repo_test.go
Comment (asraa) [reply pkg/deprecated/deprecated_repo_test.go:77]: Nice catch! this was copy-pasted. This succeeds, since the deprecated keys are added as imports.

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:39]: Hmm this was done because this assumes by the time a verifier is created (either done through unmarshalling or explcit construction) that it is well-formed. It looks like in `UnmarhsalPublicKey` that is not verified. Adding that, and adding a comment on the error handling here.

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:87]: Same as above, but in fact the entire key interface should hanlde errors in `Public()` and `PublicData()` where they currently return no errors.

File: pkg/keys/rsa.go
Comment (asraa) [reply pkg/keys/rsa.go:39]: Going to file an issue about this. It would be nice to clean this up along with the change about multiple Key IDs.

File: pkg/keys/rsa.go
Comment (asraa) [reply pkg/keys/rsa.go:84]: For now I've added checking that this will succeed when creating an rsa signer through unmarshalling, but filed https://github.com/theupdateframework/go-tuf/issues/363

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:88]: Agreed! There currently is consistency of repo key storage across impls.

File: cmd/tuf/gen_key.go
Comment (trishankatdatadog) [inline cmd/tuf/gen_key.go:14]:
```suggestion
usage: tuf gen-key [--expires=<days>] [--scheme=<scheme>] <role>
```

**Review by trishankatdatadog — COMMENTED** [review]
Thanks very much, Asra!

So just to double-check: what's the worst that can happen re:#363? We are not able to accept malformed/unusable public keys after unmarshaling, correct?

**Comment by asraa** [comment]
> So just to double-check: what's the worst that can happen re:https://github.com/theupdateframework/go-tuf/issues/363? We are not able to accept malformed/unusable public keys after unmarshaling, correct?

Good question, and I'll attach my reply to the issue as well. Currently, we use those functions after we create a `verifier` in functionality like `keys.GetVerifier`
or a `signer` through `keys.GetSigner`

https://github.com/theupdateframework/go-tuf/blob/355e39cb2df220fc3961396a6d0e30bcf2c9ac12/pkg/keys/keys.go#L57-L69

So as long as they're verified during `UnmarshalPublicKey` or `UnmarshalPrivateKey` (which now they are), then we will always have a "well-formed" enough public key to `Marshal` or get the `PublicData()` from. 

In case someone accidentally corrupts those keys in the key store or TUF metadata, then we would return errors like `tuf: error unmarshalling key:` on loading them in as `Verifier` or `Signer`. 

So worst case, yes, the key becomes un-usable on load.

That being said, it would always be good practice to re-validate the keys upon marshalling even if they were loaded in a good state, just in case. (for the issue)

**Review by trishankatdatadog — DISMISSED** [review]
Well-done, @asraa (and @toby-jn), thank you very much! Let's go 🚀

**Comment by znewman01** [comment]
ACK can look today (sorry, thought review was blocked for some reason)

File: repo.go
Comment (znewman01) [inline repo.go:315]: I'm sensing combinatorial explosion...what if I want a custom scheme but the default expiration time?

What do Go developers do in this case? (TBH I would believe you if you told me "write 2^(# options) functions 😛 )

File: pkg/keys/ecdsa.go
Comment (znewman01) [reply pkg/keys/ecdsa.go:40]: If that's the case, I'd like to see an assertion that the error is nil.

**Review by znewman01 — DISMISSED** [review]
Do you need to check "breaking change" for this PR?

I left some little comments, nice to resolve but not blocking.

My final concern (again, not blocking) has to do with deprecation. I think it's okay to immediately start producing keys using the new encoding from this library, but for repos that already exist, old clients will be around for a *long* time. Do we have a migration story there? I guess just wait until hopefully all the old clients are gone and then cut over?

**Comment by asraa** [comment]
> but for repos that already exist, old clients will be around for a long time. Do we have a migration story there? I guess just wait until hopefully all the old clients are gone and then cut over?

Right! So they can still use `pkg/deprecated` for verification, and can still construct . To migrate, they need to re-sign using the new keys, like we are doing in Sigstore. I could add the deprecated signer to `pkg/deprecated`, which would allow them to still `Sign` with the keys in their key store with the deprecated format. I think this is worth it so I'll add.

File: repo.go
Comment (asraa) [reply repo.go:315]: I think you could just use ` data.DefaultExpires(role)`? We may want to expose a `data.DefaultScheme` but I don't think it's necessary given the function sigs.

File: pkg/keys/ecdsa.go
Comment (asraa) [reply pkg/keys/ecdsa.go:40]: Since Go doesn't have assertions, I can go nuclear and use `panic(err)`: I think we should do that until the issue I filed adding graceful error handling here is resolved.

File: repo.go
Comment (znewman01) [reply repo.go:315]: Yeah, I'm wondering if maybe something like this (from Rob Pike) would be appropriate: https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html

Or maybe an options struct (number 2 [here](https://golang.cafe/blog/golang-functional-options-pattern.html)

(Totally fine to punt on this for now)

File: pkg/keys/ecdsa.go
Comment (trishankatdatadog) [reply pkg/keys/ecdsa.go:40]: Where is the 🔥 reaction emoji when you need it...

**Comment by znewman01** [comment]
> To migrate, they need to re-sign using the new keys, like we are doing in Sigstore.

Won't this immediately break every old client?

>  I could add the deprecated signer to pkg/deprecated[...]I think this is worth it so I'll add.

+1

> EDIT: I just realized that adding the deprecated signer AND verifier into the same package for compatibility is problematic for users who need to make the transition. E.g. I don't want the signer to ONLY be the deprecated version. So: the deprecated package should actually handle EITHER OR format during the transition. WDYT?

I'm not sure I'm parsing. I think there are the following possible behaviors:

For parsing:
- (default) only accept the new format
- (configurable) accept both old and new formats
- (could be an option, but I don't think we should add it) accept only the old format

This can probably be globally configurable, via the `init()` trick.

For writing new keys:
- (default) write keys in the new format
- (configurable) write keys in the old format

IMO we'd want to configure this on a key-by-key basis, rather than globally.

It'd be nice if we could do *both*, but I don't think that's possible.

**Comment by asraa** [comment]
> Won't this immediately break every old client?

Those old clients would need to import `pkg/deprecated` into their binaries. Originally I had deprecated support as a default, but then we opted to move it to an explicit package.

> +1

ECDSA Signing was never supported FYI. I just realized this as I was implementing. So I don't think it breaks old clients. They would have already implemented their own signing logic (as we did).

> (configurable) write keys in the old format

Don't think we need to opt to do this anymore, since it was never supported in the first place. go-tuf only had verification support for keys, which remains on import of `pkg/deprecated`

D: sorry this is confusing!

File: pkg/deprecated/ecdsa.go
Comment (asraa) [reply pkg/deprecated/ecdsa.go:18]: Having access to the functions without the type maps is very little use: most people want to be importing this to use in the Repo or Client logic, where all keys are pulled under the hood through these maps.

Comment (asraa) [reply pkg/deprecated/ecdsa.go:18]: Ack for testing. That makes sense I think, but I'm again not sure if it'll be used outside the tests in this package already

**Comment by trishankatdatadog** [comment]
> Those old clients would need to import `pkg/deprecated` into their binaries. Originally I had deprecated support as a default, but then we opted to move it to an explicit package.

Hold on: old clients that do not import `pkg/deprecated` will not break so long as the server keeps producing these deprecated-style public keys.

**Comment by asraa** [comment]
> Hold on: old clients that do not import pkg/deprecated will not break so long as the server keeps producing these deprecated-style public keys.

Maybe we're confused! Old clients that use old go-tuf certainly won't break because they're using an old go-tuf version. Clients that update their go-tuf dependency will break, because now ECDSA verifier expects TUF-compliant keys. Unless they import that deprecated package that understands both. RIght?

**Comment by trishankatdatadog** [comment]
> Maybe we're confused! Old clients that use old go-tuf certainly won't break because they're using an old go-tuf version. Clients that update their go-tuf dependency will break, because now ECDSA verifier expects TUF-compliant keys. Unless they import that deprecated package that understands both. RIght?

Oui. Sorry, just need to be super-clear here about what we meant about "old clients" here and so on. Thanks for the clarification!

**Comment by znewman01** [comment]
IIUC, serializing keys was never done *directly* by go-tuf. If that's true, then I am okay with 

(I'm a little confused by the discussion about signing and verifying—the signature format was always okay, right? We're just talking about signers/verifiers because that's the part of the codebase where key serialization happened?)

> > > To migrate, they need to re-sign using the new keys, like we are doing in Sigstore.
> >
> > Won't this immediately break every old client?
> 
> Those old clients would need to import `pkg/deprecated` into their binaries. Originally I had deprecated support as a default, but then we opted to move it to an explicit package.

Let me clarify: won't that immediately break every *deployed* client that's still using the old version of go-tuf?

I'm still fuzzy on the migration story. "Wait a long time before cutting over" isn't a great solution IMO; you typically don't want to break clients no matter how old.

I guess an alternative would be to replace the keys with some other key type (say, RSA for the sake of example) whose format was never affected. Then, we can distribute a new `5.root.json` with only RSA signatures (which cuts off the historical ECDSA-hex signatures, and *then* we could actually turn off the `pkg/deprecated` flag for parsing new clients. That brings the repo into compliance with the TUF spec (err, as long as you ignore everything before `5.root.json`).

File: pkg/deprecated/ecdsa.go
Comment (znewman01) [reply pkg/deprecated/ecdsa.go:18]: IMO it's worth it for testing alone. This would make it so that the tests didn't have to implement fake verifiers; I think that gives us more confidence in the cross-compatibility of the old/new versions.

Further, I think "importing this thing changes the behavior everywhere" is pretty surprising, even if it's documented, and so naming it something that makes it clearer that it's a destructive operation is a good idea.

File: pkg/deprecated/ecdsa.go
Comment (znewman01) [inline pkg/deprecated/ecdsa.go:20]: IIUC, this isn't race-y because this package depends on the package that happens to add the new verifier to the verifier map.

If that's so, then the `import` statement above is load-bearing; might be worth a comment of warning.

Alternatively (probably better) you could do a check in the other verifier map that stores only if there is no existing entry (I think that's what `LoadOrStore` does).

Either way, a comment about why this is okay would be warranted.

File: pkg/deprecated/ecdsa.go
Comment (znewman01) [inline pkg/deprecated/ecdsa.go:19]: Is this necessary? IIUC `sync.Map.Store` will clobber an existing entry (which is what we want).

**Comment by asraa** [comment]
> (I'm a little confused by the discussion about signing and verifying—the signature format was always okay, right? We're just talking about signers/verifiers because that's the part of the codebase where key serialization happened?)

Correct, signature format was always correct, it's just the way that the public key was serialized in the root metadata or how the private key was stored in the repo keystore.

> Let me clarify: won't that immediately break every deployed client that's still using the old version of go-tuf?

If they don't update their go-tuf version, they are unaffected: hex ecdsa keys will be the implementation they have in that go-tuf version. Maybe you're asking, if someone is using an old version old client, and the repository owner adds new ECDSA keys their metadata, will the old deployed client break? Yes, but that's akin to using a new feature, right? I'm not sure there's a way to get compatibility here.

> you typically don't want to break clients no matter how old.

I'd say specification deviation and lack of cross-language compatibility is one potential reason :) My feeling though is that this ONLY affected go-tuf clients who were implementing their own ECDSA signers that were compatible with the bad ECDSA verifiers in go-tuf, so they are already aware of managing this implementation.

>  Then, we can distribute a new 5.root.json with only RSA signatures (which cuts off the historical ECDSA-hex signatures, and then we could actually turn off the pkg/deprecated flag for parsing new clients. 

I see. The current plan would have already supported this, since the 5.root.json would have 5 old ECDSA signatures and 5 new ECDSA signatures. Old clients would think that the new keys are bogus, but the old keys are still there as is. There is a danger of failure on parsing though, rather than ignoring.

Because I need this PR this week, I will explore the RSA solution. I just checking in whether Go PIV tools can support RSA and some keys may

File: pkg/deprecated/ecdsa.go
Comment (asraa) [reply pkg/deprecated/ecdsa.go:19]: You're right, I don't need this, noted the import path from your other comment!

**Comment by znewman01** [comment]
Note: use `feat!:` for a breaking change

File: pkg/deprecated/ecdsa.go
Comment (asraa) [reply pkg/deprecated/ecdsa.go:20]: Testing suites mess this up, so we can't check that PEM-encoded keys are loaded at first (because each test suite will run init, and the value will be present already).

I added a comment and a check in the deprecation pkg that verifies that we DO override the only PEM verifiers in pkg/keys

File: pkg/deprecated/ecdsa.go
Comment (asraa) [reply pkg/deprecated/ecdsa.go]: Separated out the init and the package code itself: users can implement tests with a deprecated ecdsa verifier, without importing the key into the map.

**Comment by asraa** [comment]
When I have approval, I'll squash mine and Toby's commits to one and fix the commit to have `feat!`

**Comment by asraa** [comment]
Squashed, and commit message fixed, added DCO to fix Toby's old DCO issue

**Review by trishankatdatadog — DISMISSED** [review]

File: pkg/deprecated/ecdsa.go
Comment (znewman01) [reply pkg/deprecated/ecdsa.go]: Nice! I think I'd still *also* like to modify the `pkgs/keys/ecdsa.go:init()` to use `LoadOrStore` instead of `Store`, so we never accidentally replace it.

**Review by znewman01 — DISMISSED** [review]

**Review by znewman01 — APPROVED** [review]

**Review by trishankatdatadog — APPROVED** [review]

**Pull request merged**
